File: 2D Dynamic Programming/714. Best Time to Buy and Sell Stock with Transaction Fee/Solution.java
class Solution {
    private int dfs(int idx, int isBuy, int[] prices, int fee, int[][] dp){
        if (idx >= prices.length){
            return 0;
        }
        if (dp[idx][isBuy] != -1){
            return dp[idx][isBuy];
        }

        int take = Integer.MIN_VALUE;
        int notTake = dfs(idx + 1, isBuy, prices, fee, dp);
        if (isBuy == 1){
            take = dfs(idx + 1, 0, prices, fee, dp) - prices[idx];
        } else {
            take = dfs(idx + 1, 1, prices, fee, dp) + prices[idx] - fee;
        }
        dp[idx][isBuy] = Math.max(take, notTake);
        return dp[idx][isBuy];
    }
    public int maxProfit(int[] prices, int fee) {
        int n = prices.length;
        // int[][] dp = new int[n + 1][2];
        // for (int i=0; i<=n; i++){
        //     for (int j=0; j<2; j++){
        //         dp[i][j] = -1;
        //     }
        // }
        // return dfs(0, 1, prices, fee, dp);
        
        int[] prev = new int[2];
        for(int idx = n - 1; idx>=0; idx--){
            int [] curr = new int[2];
            for(int isBuy=0; isBuy<2; isBuy++){
                if (isBuy == 1){
                    int notTake = prev[isBuy];
                    int take = prev[0] - prices[idx];
                    curr[isBuy] = Math.max(take, notTake);
                } else {
                    int notTake = prev[isBuy];
                    int take = prev[1] + prices[idx] - fee;
                    curr[isBuy] = Math.max(take, notTake);
                }
            }
            prev = curr;
        }
        return prev[1];
    }
}

File: 2D Dynamic Programming/714. Best Time to Buy and Sell Stock with Transaction Fee/Solution.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        n = len(prices)
        # dp = [[-1] * (2) for _ in range(n + 1)]
        # def dfs(idx, isBuy):
        #     if idx >= n:
        #         return 0
        #     if dp[idx][isBuy] != -1:
        #         return dp[idx][isBuy]

        #     notTake = dfs(idx + 1, isBuy)

        #     if isBuy == 1:
        #         take = dfs(idx + 1, 0) - prices[idx]
        #     elif isBuy == 0:
        #         take = dfs(idx + 1, 1) + prices[idx] - fee

        #     dp[idx][isBuy] = max(notTake, take)
        #     return dp[idx][isBuy]
        
        # return dfs(0, 1)

        prev = [0] * 2
        for idx in range(n - 1, -1, -1):
            curr = [0] * 2
            for isBuy in range(2):
                if isBuy == 1:
                    notTake = prev[1]
                    take = prev[0] - prices[idx]
                    curr[isBuy] = max(notTake, take)
                else:
                    notTake = prev[0]
                    take = prev[1] + prices[idx] - fee
                    curr[isBuy] = max(notTake, take)
            prev = curr
        return prev[1]